src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java

```diff
@@ -99,22 +99,44 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
         // TODO it might be sensible to consolidate this with the overridden method in
         // order to minimize duplicated logic
 
-        final SecurityDynamicConfiguration<?> configuration = load(getConfigName(), false);
+        final SecurityDynamicConfiguration<?> internalUsersConfiguration = load(getConfigName(), false);
 
-        if (isHidden(configuration, username)) {
+        if (isHidden(internalUsersConfiguration, username)) {
             forbidden(channel, "Resource '" + username + "' is not available.");
             return;
         }
 
         // check if resource is writeable
-        if (!isReservedAndAccessible(configuration, username)) {
+        if (!isReservedAndAccessible(internalUsersConfiguration, username)) {
             forbidden(channel, "Resource '" + username + "' is read-only.");
             return;
         }
 
         final ObjectNode contentAsNode = (ObjectNode) content;
         final SecurityJsonNode securityJsonNode = new SecurityJsonNode(contentAsNode);
 
+        // Don't allow user to add hidden, reserved or non-existent role
+        final List<String> opendistroSecurityRoles = securityJsonNode.get("opendistro_security_roles").asList();
+        if (opendistroSecurityRoles != null) {
+            final SecurityDynamicConfiguration<?> rolesConfiguration = load(CType.ROLES, false);
+            for (final String role: opendistroSecurityRoles) {
+
+                if (rolesConfiguration.getCEntry(role) == null) {
+                    notFound(channel, "Role '"+role+"' does not exist.");
+                }
+
+                if (isHidden(rolesConfiguration, role)) {
```

Comment:
move `if (rolesConfiguration.getCEntry(role) == null)` to here:
```suggestion
                if (rolesConfiguration.exists(role) || isHidden(rolesConfiguration, role)) {
```

Reply:
I am slightly conflicted about treating hidden and non-existent roles the same way. 
For non-existent roles the customer can resolve the issue by creating the new role, for hidden roles this will not be possible. If we return the same error message for both do you think it will cause confusion for the customer?

Reply:
Hidden resource(role) means that for all practical purposes it can't be distinguished from not existing resource(role). Check, for example, how patch request for hidden resource/role is handled. The only exception where it is not possible to avoid exposing hidden resources is create(put) request.